File: src/boj/2437.go
package main

import (
	"bufio"
	"os"
	"sort"
	"strconv"
)

var (
	sc *bufio.Scanner
	wr *bufio.Writer
)

func init() {
	sc = bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)
	wr = bufio.NewWriter(os.Stdout)
}

func scanInt() int {
	sc.Scan()
	n, _ := strconv.Atoi(sc.Text())
	return n
}

func main() {
	defer wr.Flush()

	var N = scanInt()
	var nums = make([]int, N)
	for i := 0; i < N; i++ {
		nums[i] = scanInt()
	}
	sort.Ints(nums)

	wr.WriteString(strconv.Itoa(rec(nums, 1, 0)))
}

func rec(arr []int, next, k int) int {
	if k >= len(arr) || arr[k] > next {
		return next
	}

	return rec(arr, next+arr[k], k+1)
}

File: src/boj/2437.js
const input = () => {
  return require("fs")
    .readFileSync(0, "utf8")
    .trim()
    .split("\n")
    .pop()
    .split(" ")
    .map(Number);
};

const compute = (nums) => {
  nums.sort((a, b) => a - b);

  const arr = [-1];

  const rec2 = (level) => {
    if (level === nums.length) {
      console.log(arr);
      return;
    }

    for (let i = 0; i < arr.length; i++) {
      if (arr[i] == null) continue;
      if (arr[i] !== level - 1) continue;

      for (const n of nums) {
        arr[i + n] = level;
      }
    }

    console.log(arr);

    rec2(level + 1);
  };

  rec2(0);

  // const rec = (i, cases) => {
  //   if (i === nums.length) {
  //     return cases.length;
  //   }

  //   const next = [...new Set([...cases, ...cases.map((n) => n + nums[i])])];

  //   if (next[next.length - 1] > next.length - 1) {
  //     const j = next.findIndex((n, i) => n !== i) - 1;
  //     return next[j] + 1;
  //   }

  //   return rec(i + 1, next);
  // };

  // return rec(0, [0]);
};

console.log(compute(input()));
